scripts/utils.py

```diff
@@ -337,3 +344,403 @@ def add_population_to_dataframe(
+        # Count the number of missing values in each group.
+        num_nans_detected = df.groupby(groupby_columns, dropna=False).agg(
+            lambda x: pd.isnull(x).sum()
+        )
+        # Make nan any aggregation where there were too many missing values.
+        grouped = grouped[num_nans_detected <= num_allowed_nans]
+
+    if frac_allowed_nans is not None:
+        # Count the number of missing values in each group.
+        num_nans_detected = df.groupby(groupby_columns, dropna=False).agg(
+            lambda x: pd.isnull(x).sum()
+        )
+        # Count number of elements in each group (avoid using 'count' method, which ignores nans).
+        num_elements = df.groupby(groupby_columns, dropna=False).size()
+        # Make nan any aggregation where there were too many missing values.
+        grouped = grouped[
+            num_nans_detected.divide(num_elements, axis="index") <= frac_allowed_nans
+        ]
+
+    return grouped
+
+
+class RegionNotFound(ExceptionFromDocstring):
+    """Region was not found in countries-regions dataset."""
+
+
+def list_countries_in_region(region, countries_regions=None, income_groups=None):
+    """List countries that are members of a region.
+
+    Parameters
+    ----------
+    region : str
+        Name of the region (e.g. Europe).
+    countries_regions : pd.DataFrame or None
+        Countries-regions dataset, or None to load it from the catalog.
+
+    Returns
+    -------
+    members : list
+        Names of countries that are members of the region.
+
+    """
+    # TODO: Add tests.
+    if countries_regions is None:
+        countries_regions = catalog.find(
+            "countries_regions", dataset="reference", namespace="owid"
+        ).load()
+
+    if income_groups is None:
+        income_groups = catalog.find(
+            table="wb_income_group", dataset="wb_income", namespace="wb"
+        )
+        income_groups = (
+            income_groups[income_groups.path.str.startswith("garden")]
+            .load()
+            .reset_index()
+        )
+    income_groups_names = income_groups["income_group"].dropna().unique().tolist()
+
+    # TODO: Once countries-regions has additional columns 'is_historic' and 'is_country', select only countries, and not
+    #  historical regions.
+
+    if region in countries_regions["name"].tolist():
+        # Find codes of member countries in this region.
+        member_codes = json.loads(
+            countries_regions[countries_regions["name"] == region]["members"].item()
+        )
+        # Get standardized names of these countries.
+        members = countries_regions.loc[member_codes]["name"].tolist()
+    elif region in income_groups_names:
+        members = (
+            income_groups[income_groups["income_group"] == region]["country"]
+            .unique()
+            .tolist()
+        )
+    else:
+        raise RegionNotFound
+
+    return members
+
+
+def list_countries_in_region_with_largest_contribution(
```

Comment:
Nitpick - if you only plan to use the function internally (i.e. users of this library won't be exporting it often), it's nice to prefix it with `_`. That makes it clear what functions are meant to be used and what aren't. You also don't need so detailed docs for them (e.g. not duplicating parameter docs that are already in the function calling this one). Up to you though!